File: src/cmdbox/resolve/reference_parsing.py
from typing import Optional

from .type_defs import RefKind


def read_angle_token(s: str, start_i: int) -> tuple[Optional[str], str, int]:
    """
    Reads a token beginning with "<" until an unescaped ">" is encountered.

    Args:
        s (str): The string containing the token.
        start_i (int): The index of the first character of the token.

    Returns:
        str: token_inner (with escapes already interpreted), or None if no closing bracket.
        str: raw_token - the original token string, including the angle brackets.
        int: next_i - the index of the first character after the closing bracket.

    """
    assert s[start_i] == "<"
    i = start_i + 1
    n = len(s)

    inner_chars: list[str] = []

    while i < n:
        ch = s[i]
        if ch == "\\":
            if i + 1 < n and s[i + 1] in ("\\", "<", ">"):
                inner_chars.append(s[i + 1])
                i += 2
                continue
            inner_chars.append("\\")
            i += 1
            continue

        if ch == ">":
            raw_token = s[start_i + 1 : i]
            token_inner = "".join(inner_chars).strip()
            return token_inner, raw_token, i + 1

        inner_chars.append(ch)
        i += 1

    return None, s[start_i:], n


def parse_kind_and_key(token_inner: str) -> tuple[RefKind, str]:
    """
    Parses the kind and key from a given token string.

    This method evaluates the provided token and determines its type, which can be a
    command or variable, based on the prefix present in the token. If no prefix is
    found, it defaults to `RefKind.VARIABLE`. The method ensures that parts of the
    token are appropriately stripped of whitespace during processing.

    Args:
        token_inner (str): The input token string that may specify a prefix and a key,
            separated by a colon.

    Returns:
        tuple[RefKind, str]: A tuple containing the kind of reference (`RefKind`) and
        its corresponding key as a string.
    """
    if ":" not in token_inner:
        return RefKind.VARIABLE, token_inner
    prefix, key = (part.strip() for part in token_inner.split(":", 1))

    if prefix == "cmd":
        return RefKind.COMMAND, key
    if prefix == "var":
        return RefKind.VARIABLE, key
    return RefKind.VARIABLE, token_inner


def extract_references(template: str) -> list[tuple[RefKind, str]]:
    """
    Extracts references from a given template string.

    Parses through the input template string to identify specific patterns enclosed
    in angle brackets (`<` and `>`). Each such identified token is processed to
    extract its kind and key, which are then appended to a list of references
    represented as tuples.

    Args:
        template (str): The input string containing the template to be parsed.

    Returns:
        list[tuple[RefKind, str]]: A list of tuples where each tuple contains the
        kind and key extracted from the references found within the template.
    """
    refs: list[tuple[RefKind, str]] = []
    i = 0
    n = len(template)

    while i < n:
        ch = template[i]
        if ch == "\\":
            i += 2
            continue
        if ch == "<":
            token_inner, _, next_i = read_angle_token(template, i)
            if token_inner:
                refs.append(parse_kind_and_key(token_inner))
            i = next_i
            continue
        i += 1

    return refs

File: internal/resolve/reference_parsing.go
package resolve

import (
	"fmt"
	"strings"
)

// ReadAngleToken parses a substring starting at the provided index to extract an angle bracket token and its inner content.
// It returns a pointer to the trimmed inner content, the raw token string, and the index after the closing '>'.
// Panics if the substring at startIndex does not begin with '<'.
//
// ReadAngleToken panics if s[startIndex] is not '<'. Callers must confirm this themselves before calling.
func ReadAngleToken(runes []rune, startIndex int) (*string, string, int) {
	n := len(runes)

	if runes[startIndex] != '<' {
		panic(fmt.Errorf("expected '<' at index %d", startIndex))
	}

	i := startIndex + 1
	var innerChars []rune

	for i < n {
		ch := runes[i]
		if ch == '\\' {
			if i+1 < n {
				next := runes[i+1]
				if next == '\\' || next == '<' || next == '>' {
					innerChars = append(innerChars, next)
					i += 2
					continue
				}
			}
			innerChars = append(innerChars, '\\')
			i++
			continue
		}
		if ch == '>' {
			rawToken := string(runes[startIndex+1 : i])
			tokenInner := strings.TrimSpace(string(innerChars))
			return &tokenInner, rawToken, i + 1
		}
		innerChars = append(innerChars, ch)
		i++
	}
	return nil, string(runes[startIndex:]), n
}

// ParseKindAndKey extracts the reference kind and key from a token string in the format "prefix:key" or "key".
func ParseKindAndKey(tokenInner string) (RefKind, string) {
	if !strings.Contains(tokenInner, ":") {
		return RefKindVariable, tokenInner
	}

	parts := strings.SplitN(tokenInner, ":", 2)
	prefix := strings.TrimSpace(parts[0])
	key := strings.TrimSpace(parts[1])

	if prefix == "cmd" {
		return RefKindCommand, key
	}
	if prefix == "var" {
		return RefKindVariable, key
	}
	return RefKindVariable, tokenInner
}

type Ref struct {
	Kind RefKind
	Key  string
}

// ExtractReferences scans the provided template string, identifies and extracts all references enclosed in angle brackets.
// It returns a slice of Ref structs, each containing the kind and key derived from a reference token.
func ExtractReferences(template string) []Ref {
	runes := []rune(template)
	var refs []Ref
	i := 0
	n := len(runes)

	for i < n {
		ch := runes[i]
		if ch == '\\' {
			i += 2
			continue
		}
		if ch == '<' {
			tokenInner, _, nextIdx := ReadAngleToken(runes, i)
			if tokenInner != nil && *tokenInner != "" {
				kind, key := ParseKindAndKey(*tokenInner)
				refs = append(refs, Ref{Kind: kind, Key: key})
			}
			i = nextIdx
			continue
		}
		i++
	}
	return refs
}
